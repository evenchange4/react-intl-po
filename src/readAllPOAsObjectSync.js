import { sync as globSync } from 'glob';
import path from 'path';
import po2json from 'po2json';
import mapValues from 'lodash/mapValues';
import toObjectBy from 'to-object-by';

export const DEFAULT_MAPPER = filepath =>
  path.basename(filepath).match(/([^.]*\.)*([^.]+)\.po$/)[2];

/**
 * Read translated .po file synchronized and
 * aggregates translated messages object
 *
 * @param {String} srcPatterns - path to translated .po file
 * @return {Object} po - return aggregates object
 *
 * @author Michael Hsu
 */

function readAllPOAsObjectSync(srcPatterns, localeMapper = DEFAULT_MAPPER) {
  const filepaths = globSync(srcPatterns);

  return toObjectBy(filepaths, filepath => {
    const json = po2json.parseFileSync(filepath);
    const translated = mapValues(json, o => o[1]); // omit plural
    const locale = localeMapper(filepath); // parse locale name

    return {
      [locale]: translated,
    };
  });
}

export default readAllPOAsObjectSync;
