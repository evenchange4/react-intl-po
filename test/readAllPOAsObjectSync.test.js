import readAllPOAsObjectSync, {
  DEFAULT_MAPPER as defaultMapper,
} from '../src/readAllPOAsObjectSync';

it('should return a function', () => {
  expect(typeof readAllPOAsObjectSync).toBe('function');
});

it('should return locale when DEFAULT_MAPPER w/o prefix', () => {
  expect(defaultMapper('mcs-public.zh-CN.po')).toBe('zh-CN');
  expect(defaultMapper('mcs-public.zh-TW.po')).toBe('zh-TW');
  expect(defaultMapper('mcs-public.en.po')).toBe('en');

  expect(
    defaultMapper('./node_modules/mcs-translation/po/mcs-public.zh-cn.po'),
  ).toBe('zh-cn');
  expect(
    defaultMapper('./node_modules/mcs-translation/po/mcs-public.zh-tw.po'),
  ).toBe('zh-tw');
  expect(
    defaultMapper('./node_modules/mcs-translation/po/mcs-public.en.po'),
  ).toBe('en');

  expect(defaultMapper('./node_modules/mcs-translation/po/zh-CN.po')).toBe(
    'zh-CN',
  );
  expect(defaultMapper('./node_modules/mcs-translation/po/zh-TW.po')).toBe(
    'zh-TW',
  );
  expect(defaultMapper('./node_modules/mcs-translation/po/en.po')).toBe('en');

  expect(defaultMapper('zh-CN.po')).toBe('zh-CN');
  expect(defaultMapper('zh-TW.po')).toBe('zh-TW');
  expect(defaultMapper('en.po')).toBe('en');
});

it('should return po object with default localeMapper', () => {
  expect(readAllPOAsObjectSync('./test/po/mcs-public.*.po')).toMatchSnapshot();
});

it('should return po object with custom localeMapper', () => {
  expect(
    readAllPOAsObjectSync(
      './test/po/*_project.po',
      filename => filename.split(/(\/|_)/g)[6],
    ),
  ).toMatchSnapshot();
});
