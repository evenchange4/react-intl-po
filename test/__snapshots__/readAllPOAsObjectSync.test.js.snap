// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return po object with custom localeMapper 1`] = `
{
  "zh-CN": {
    "( default )": "（默认值）"
  },
  "zh-TW": {
    "( default )": "（默認值）"
  }
}
`;

exports[`should return po object with default localeMapper 1`] = `
{
  "zh-CN": {
    "Creator": "建立者（簡中）",
    "Version": "版本（簡中）"
  },
  "zh-TW": {
    "Creator": "建立者",
    "Version": "版本"
  }
}
`;
